server/lib/parties/party-service.ts

```diff
@@ -0,0 +1,217 @@
+          from: party!.leader,
+        }),
+        () => {
+          // TODO(2Pac): Handle user quitting; need to keep a map of user -> invites?
+        },
+      )
+    })
+  }
+
+  removeInvite(partyId: string, target: PartyUser, leader?: PartyUser) {
+    const party = this.parties.get(partyId)
+    if (!party) {
+      throw new PartyServiceError(ErrorEnum.NotFound, 'Party not found')
+    }
+
+    if (leader && leader.id !== party.leader) {
+      throw new PartyServiceError(ErrorEnum.Forbidden, 'Must be party leader')
+    }
+
+    const updatedParty = party.deleteIn(['invites', target.id])
+    this.parties = this.parties.set(party.id as string, updatedParty)
+
+    const targetUser = this.getUser(target.name as string)
+    targetUser.unsubscribe(getInvitesPath(partyId))
+
+    this.publishToParty(partyId, {
+      type: 'decline',
+      target,
+    })
+
+    // TODO(2Pac): Do we want to remove the party record if there's only one member left in a party
+    // and no outstanding invites?
+  }
+
+  acceptInvite(partyId: string, user: PartyUser, clientId: string) {
+    const client = this.getClient(user.id as number, clientId)
+
+    const party = this.parties.get(partyId)
+    if (!party) {
+      throw new PartyServiceError(ErrorEnum.NotFound, 'Party not found')
+    }
+
+    if (party.members.size > MAX_PARTY_SIZE) {
+      throw new PartyServiceError(ErrorEnum.Conflict, 'Party is full')
+    }
+
+    const oldParty = this.getClientParty(client)
+    if (oldParty) {
+      // TODO(2Pac): Handle switching parties
+    }
+
+    const updatedParty = party.deleteIn(['invites', user.id]).setIn(['members', user.id], user)
+    this.parties = this.parties.set(party.id as string, updatedParty)
+    this.clientToPartyId = this.clientToPartyId.set(client, partyId)
+
+    this.publishToParty(partyId, {
+      type: 'join',
+      user,
+    })
+
+    const userSockets = this.getUser(user.name as string)
+    userSockets.unsubscribe(getInvitesPath(partyId))
+    this.subscribeToParty(client, updatedParty)
+  }
+
+  private subscribeToParty(client: ClientSocketsGroup, party: PartyRecord) {
+    client.subscribe(
+      getPartyPath(party.id as string),
+      () => ({
+        action: 'init',
+        party,
+      }),
+      () => this.handleClientQuit(client),
+    )
+  }
+
+  private handleClientQuit(client: ClientSocketsGroup) {
+    const party = this.getClientParty(client)
+    if (!party) {
+      throw new PartyServiceError(ErrorEnum.NotFound, 'Party not found')
+    }
+
+    this.clientToPartyId.delete(client)
+    this.parties = this.parties.deleteIn([party.id, 'members', client.userId])
+    this.publishToParty(party.id as string, {
+      type: 'leave',
+      user: client.userId,
+    })
+
+    // TODO(2Pac): Handle party leader leaving
+
+    // TODO(2Pac): Handle last person in a party leaving
+  }
+
+  private publishToParty(partyId: string, data: any) {
+    this.nydus.publish(getPartyPath(partyId as string), data)
```

Comment:
Not really sure what this cast is here for, this is already the type of this parameter